fendermint/app/src/metrics/tracing.rs

```diff
@@ -62,28 +66,28 @@ macro_rules! inc_counter {
     };
 }
 
+/// Produce the prefixed event name from the type name.
 macro_rules! event_name {
     ($event_ty:ident) => {
         concat!("event::", stringify!($event_ty))
     };
 }
 
+/// Call one of the macros that set values on a metric.
 macro_rules! event_mapping {
-    (gauges, $event:ident, $event_ty:ident :: $field:ident, $metric:expr) => {
-        set_gauge!($event, $event_ty::$field, $metric);
-    };
-    (counters, $event:ident, $event_ty:ident :: $field:ident, $metric:expr) => {
-        inc_counter!($event, $event_ty::$field, $metric);
+    ($op:ident, $event:ident, $event_ty:ident :: $field:ident, $metric:expr) => {
+        $op!($event, $event_ty::$field, $metric);
     };
 }
 
+/// Match the event name to event DTO types and within the map fields to metrics.
 macro_rules! event_match {
```

Comment:
I find this `event_match` and `event_mapping` macros quite complex for what its doing. Why can't we just keep it simple and have some metrics source file that contains all our static global metrics and then a register function somewhere where we initialize fendermint?

Reply:
> Why can't we just keep it simple and have some metrics source file that contains all our static global metrics

This is exactly what `prometheus.rs` is.

>  and then a register function somewhere where we initialize fendermint?

This is a bit more vague. What does your "register function" do? 

Note that there is a register function is `prometheus.rs` and it's indeed called during the initialisation of fendermint.

Reply:
So, I can imagine that the `fendermint_tracing` crate would have a macro for _defining_ events, which would look very similar to what you see below, and it would define the event DTO as well as the metrics at the same time, with a `register_metrics` to register all metrics in that source module (which one has to remember to call, unless they register into the default Prometheus registry, which according to `prometheus_exporter` might be different than what it itself is using). 

Every defined event would then have its own `emit` method that does 1) logging and 2) increment metrics, so there wouldn't be a need for the subscription to the emitted logs for the purpose of metrics any more. 

It would be a different way of achieving the same result:
* metrics in one file vs co-located with the events with multiple registration functions across the codebase
* direct metrics manipulation vs subscription (note that the current is closer to Promtail, but type safe)
* easier to define different log levels per event (we can do it today as well, just need to modify the subscription to ignore the level)
* every producer crate depends on Prometheus, rather than just tracing
* when reused as a library, crates come with their metrics, no need for subscribing/mapping/tailing
* dilutes the concept of events, one step closer to logging and metrics everywhere
* more difficult to list all the metrics in one place
* more difficult to see the source of data for metrics in once place
* even less room for errors like incrementing the wrong gauge
* more difficult to control the naming of metrics

Reply:
Ok, you convinced me, this sounds reasonable to me!